### DISCUSSION TITLE
how to implement msm for element exceeding MODULUS in ecc.BN254.ScalarField()

### DISCUSSION THREAD

**Discussion by xiangjianmeng** [Consensys/gnark#1186] [General]
The full example is the Unit test `TestMsmSolve` in https://github.com/xiangjianmeng/gnark-halo2-verify/tree/fix-compile

**Comment by ivokub** [comment]
Instead of using `ToElement`, you can provide the emulated elements as witness to the circuit and then initialize then during witness assignment using `emulated.ValueOf`.

So instead of:
```go
type yourCircuit struct {
    X frontend.Variable
}

func (c *yourCircuit) Define(api frontend.API) error {
    xnna, err := ToElement(c.X)
    // xnna is type emulated.Element[emulated.BN254Fp]
   // ...
}
```
you could instead have
```go
type nnaCircuit struct {
    X emulated.Element[emulated.BN254Fp]
}

func (c *nnaCircuit) Define(api frontend.API) error {
    // use c.X as emulated.Element[emulated.BN254Fp] directly
   // ...
}
```

This approach works similarly for providing the point directly as witness element, see for example https://github.com/Consensys/gnark/blob/master/std/algebra/emulated/sw_emulated/point_test.go#L622-L661

**Reply by xiangjianmeng** [reply]
What is the normal way to implement business logic using gnark circuit:
1. implement business logic using gnark circuit derectly, I need to use the result caclulated in circuit like hash, then as scalar to caculate msm.
2. or define all immediate results in circuit, firstly implement business logic just using go, then assign immediate results to witness. secondly implement circuit to check witness which satisfies constraints.

**Reply by ivokub** [reply]
The second approach, you use a SNARK circuit to provide a proof that some statement is correct. But you still need to compute the statement externally.

For example - SNARK circuits encode "there exists a signature satisfying some conditions" or "there exists a message which hashes to digest" etc, but you have to know the signature and the prehash etc.

**Reply by xiangjianmeng** [reply]
@ivokub call verifyProof of generated contract_groth16_test.sol error: ProofInvalid : The proof is invalid.
here is the full example, thanks a lot:
```
type TestBN254ScalarMul struct {
	Point  sw_bn254.G1Affine
	Scalar sw_bn254.Scalar
	Res    sw_bn254.G1Affine `gnark:",public"`
}

func (c *TestBN254ScalarMul) Define(api frontend.API) error {
	cr, err := sw_emulated.New[emparams.BN254Fp, emparams.BN254Fr](api, sw_emulated.GetBN254Params())
	if err != nil {
		return err
	}
	res := cr.ScalarMul(&c.Point, &c.Scalar)
	cr.AssertIsEqual(res, &c.Res)
	return nil
}

func TestMsmSolve(t *testing.T) {
	assert := test.NewAssert(t)
	x, _ := new(big.Int).SetString("1", 10)
	y, _ := new(big.Int).SetString("21888242871839275222246405745257275088696311157297823662689037894645226208581", 10)
	scalar, _ := new(big.Int).SetString("21147276235438245106538451154094232271190030085887596632745409482267565260819", 10)

	//var blob []byte
	//blob = append(blob, x.FillBytes(make([]byte, 32))...)
	//blob = append(blob, y.FillBytes(make([]byte, 32))...)
	//
	//p := new(bn256.G1)
	//_, err := p.Unmarshal(blob)
	//assert.NoError(err)
	//
	//res := new(bn256.G1)
	//res.ScalarMult(p, scalar)
	//xStr, yStr, _ := extractAndConvert(res.String())
	//var resCircuit bn254.G1Affine
	//_, err = resCircuit.X.SetString(xStr)
	//assert.NoError(err)
	//_, err = resCircuit.Y.SetString(yStr)
	//assert.NoError(err)

	//xStr := "1"
	//yStr := "21888242871839275222246405745257275088696311157297823662689037894645226208581"
	//scaStr := "21147276235438245106538451154094232271190030085887596632745409482267565260819"

	//xEle, _ := new(fp.Element).SetString(xStr)
	//yEle, _ := new(fp.Element).SetString(yStr)

	var point = bn254.G1Affine{
		X: *new(fp.Element).SetBigInt(x),
		Y: *new(fp.Element).SetBigInt(y),
	}
	assert.True(point.IsOnCurve())
	var res1 bn254.G1Affine
	res1.ScalarMultiplication(&point, scalar)
	//assert.Equal(resCircuit, res1)

	//var u fr.Element
	//scalarEle := u.SetBigInt(scalar)

	witnessCircuit := TestBN254ScalarMul{
		Point: sw_bn254.G1Affine{
			X: emulated.ValueOf[emulated.BN254Fp](x),
			Y: emulated.ValueOf[emulated.BN254Fp](y),
		},
		Scalar: emulated.ValueOf[emulated.BN254Fr](scalar),
		Res: sw_bn254.G1Affine{
			X: emulated.ValueOf[emulated.BN254Fp](res1.X),
			Y: emulated.ValueOf[emulated.BN254Fp](res1.Y),
		},
	}
	circuit := TestBN254ScalarMul{}

	err := test.IsSolved(&circuit, &witnessCircuit, ecc.BN254.ScalarField())
	assert.NoError(err)

	r1cs, err := frontend.Compile(ecc.BN254.ScalarField(), r1cs.NewBuilder, &circuit)
	if err != nil {
		log.Fatalf("Failed to compile circuit: %v", err)
	}

	pk, vk, err := groth16.Setup(r1cs)
	if err != nil {
		log.Fatalf("Failed to setup keys: %v", err)
	}

	witness, err := frontend.NewWitness(&witnessCircuit, ecc.BN254.ScalarField())
	if err != nil {
		panic(err)
	}

	proof, err := groth16.Prove(r1cs, pk, witness)
	if err != nil {
		log.Fatalf("Failed to create proof: %v", err)
	}
	proofJSON, _ := json.MarshalIndent(proof, "", "    ")
	_ = os.WriteFile("gnark_proof_test.json", proofJSON, 0644)
	fProof, err := os.Create("proof_test")
	if err != nil {
		log.Fatalln(err)
	}
	_, err = proof.WriteRawTo(fProof)
	if err != nil {
		log.Fatalln(err)
	}

	public, err := witness.Public()
	if err != nil {
		log.Fatalf("Failed to Public: %v", err)
	}
	s, err := frontend.NewSchema(&witnessCircuit)
	if err != nil {
		panic(err)
	}
	publicWitnessJSON, err := public.ToJSON(s)
	if err != nil {
		panic(err)
	}
	_ = os.WriteFile("gnark_inputs_test.json", publicWitnessJSON, 0644)
	fPublic, err := os.Create("public_test")
	if err != nil {
		log.Fatalln(err)
	}
	_, err = public.WriteTo(fPublic)
	if err != nil {
		log.Fatalln(err)
	}

	if err := groth16.Verify(proof, vk, public); err != nil {
		log.Fatalf("Failed to verify proof: %v", err)
	}

	f, err := os.Create("contract_groth16_test.sol")
	if err != nil {
		log.Fatalln(err)
	}
	err = vk.ExportSolidity(f)
	if err != nil {
		log.Fatalln(err)
	}

	//err = test.IsSolved(&circuit, &witnessCircuit, ecc.BN254.ScalarField())
	assert.NoError(err)
}
```
- gnark_proof_test.json
```
{
    "Ar": {
        "X": "4144358347204077126778200311321555161723370155532307701659521933515231823162",
        "Y": "14451125246523983184981498811141315864322677539714352184777030186995598755535"
    },
    "Krs": {
        "X": "968432878662322189009767591374284349353669919292564120694774737407779918332",
        "Y": "6701101671904409882176170683356539607954808749554916968162525713180769339731"
    },
    "Bs": {
        "X": {
            "A0": "3066257380253317487685256334290279507505189384070208107440660339279124207704",
            "A1": "2914679704278130366768703119377893284928343414280615124860891141219373649715"
        },
        "Y": {
            "A0": "9223874091757770385713568403393878908357520127372589131762178350953338840018",
            "A1": "19262378583736507933941203720910403603220198635510074358879271357762984850986"
        }
    },
    "Commitments": [
        {
            "X": "1799949527379213147398034849495234585189838872454242258355524026130131131726",
            "Y": "5877455121919277049556529789683122391600146752920534205851491108107694099985"
        }
    ],
    "CommitmentPok": {
        "X": "9892429236022098093312078149843514141171168416812907815645774631378689022362",
        "Y": "1224695008229435748109336636519411864986889080338770518464751084634149169660"
    }
}
```

- gnark_input_test.json
```
{"Proof":[null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null],"VerifyInst":[null],"Aux":[null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null],"TargetInst":["7059793422771910484","2556686405730241944","2133554817341762742","8974371243071329347"]}
```

- call data
```
uint256[8] calldata proof: [19701848315313497362967702456328072430342313124619018744636938522161947405995,1669883390735264312810852248375644707077053561256874780573178089353083407741,8000016743751859809583536455088953876615317704138756160351838790104827818547,12796829509417226163618615292337795673921540177061027891408970257672954921359,2626075933215950069690247983365518351432916415443922967058411018556451341852,5441809839697931038127731382042871881069187609171798889382506497091095906431,16835649715159260952518951233487805707721048256699866258448410229155687049993,12438345056061558306791268454539078073205210470460804459347867631713356855604]
uint256[2] calldata commitments: [18633562070859979286009296839104602914722384776624865698342983343993354853301,19700056948682709433783414448710695829094473364144285027012889738182867985092]
uint256[2] calldata commitmentPok: [7861280587763793072055822964162760191633562213779903923080044200166331613628,17043649752569401864415461940499547535984073633061985147517904248363141174404]
uint256[8] calldata input: [16404902109909450017,2800271095994586453,14313083841630603949,959314423801859314,12423662052074672397,9788095319943472193,7353901976489907706,2661334625146839976]
```

**Reply by ivokub** [reply]
One option to run Solidity test is by using the `assert.Checkcircuit` method:
```go
package issue1186

import (
	"math/big"
	"testing"

	"github.com/consensys/gnark-crypto/ecc"
	"github.com/consensys/gnark-crypto/ecc/bn254"
	"github.com/consensys/gnark-crypto/ecc/bn254/fp"
	"github.com/consensys/gnark/backend"
	"github.com/consensys/gnark/frontend"
	"github.com/consensys/gnark/std/algebra/emulated/sw_bn254"
	"github.com/consensys/gnark/std/algebra/emulated/sw_emulated"
	"github.com/consensys/gnark/std/math/emulated"
	"github.com/consensys/gnark/std/math/emulated/emparams"
	"github.com/consensys/gnark/test"
)

type TestBN254ScalarMul struct {
	Point  sw_bn254.G1Affine
	Scalar sw_bn254.Scalar
	Res    sw_bn254.G1Affine `gnark:",public"`
}

func (c *TestBN254ScalarMul) Define(api frontend.API) error {
	cr, err := sw_emulated.New[emparams.BN254Fp, emparams.BN254Fr](api, sw_emulated.GetBN254Params())
	if err != nil {
		return err
	}
	res := cr.ScalarMul(&c.Point, &c.Scalar)
	cr.AssertIsEqual(res, &c.Res)
	return nil
}

func TestMsmSolve(t *testing.T) {
	assert := test.NewAssert(t)
	x, _ := new(big.Int).SetString("1", 10)
	y, _ := new(big.Int).SetString("21888242871839275222246405745257275088696311157297823662689037894645226208581", 10)
	scalar, _ := new(big.Int).SetString("21147276235438245106538451154094232271190030085887596632745409482267565260819", 10)

	var point = bn254.G1Affine{
		X: *new(fp.Element).SetBigInt(x),
		Y: *new(fp.Element).SetBigInt(y),
	}
	assert.True(point.IsOnCurve())
	var res1 bn254.G1Affine
	res1.ScalarMultiplication(&point, scalar)

	witnessCircuit := TestBN254ScalarMul{
		Point: sw_bn254.G1Affine{
			X: emulated.ValueOf[emulated.BN254Fp](x),
			Y: emulated.ValueOf[emulated.BN254Fp](y),
		},
		Scalar: emulated.ValueOf[emulated.BN254Fr](scalar),
		Res: sw_bn254.G1Affine{
			X: emulated.ValueOf[emulated.BN254Fp](res1.X),
			Y: emulated.ValueOf[emulated.BN254Fp](res1.Y),
		},
	}
	circuit := TestBN254ScalarMul{}

	assert.CheckCircuit(&circuit, test.WithValidAssignment(&witnessCircuit), test.WithBackends(backend.GROTH16), test.WithCurves(ecc.BN254))
}
```

We don't run solidity checker by default, to ensure it is run, use the `solccheck` build tag:
```bash
go test -timeout 10m -tags solccheck,prover_checks -run ^TestMsmSolve$ github.com/consensys/gnark/examples/issue1186 -v -count=1
```

For this to work, you need to have an additional utility `gnark-solidity-checker`. We have it in a separate repository to avoid huge dependency list introduced by importing `go-ethereum`. See https://github.com/consensys/gnark-solidity-checker for how to install its dependencies and then run
```bash
go install github.com/consensys/gnark-solidity-checker@master
```

In that repository you can also see how to construct the inputs for the Groth16 verifier i.e. https://github.com/Consensys/gnark-solidity-checker/blob/main/cmd/templates.go#L72-L124. We don't have a good way to export the full proof as a byte slice yet as in PLONK solidity verifier, so you need to do it a bit manually, but hopefully the example is clear. We have an open issue for that https://github.com/Consensys/gnark/issues/956.

Sorry for the difficult setup, we are now mostly focused on supporting PLONK Solidity verifier, but will try to improve Groth16 verifier also as it is still quite much used by the community.

**Reply by ivokub** [reply]
But for running the `assert.CheckCircuit`, I discovered a bug which is fixed here https://github.com/Consensys/gnark/pull/1187

**Reply by xiangjianmeng** [reply]
I have executed `assert.CheckCircuit` like this command successfully, but still verifyProof on sepolia failed.
```
go test -timeout 10m -tags solccheck,prover_checks -run TestMsmSolve
```

**Reply by xiangjianmeng** [reply]
> One option to run Solidity test is by using the `assert.Checkcircuit` method:
> 
> ```go
> package issue1186
> 
> import (
> 	"math/big"
> 	"testing"
> 
> 	"github.com/consensys/gnark-crypto/ecc"
> 	"github.com/consensys/gnark-crypto/ecc/bn254"
> 	"github.com/consensys/gnark-crypto/ecc/bn254/fp"
> 	"github.com/consensys/gnark/backend"
> 	"github.com/consensys/gnark/frontend"
> 	"github.com/consensys/gnark/std/algebra/emulated/sw_bn254"
> 	"github.com/consensys/gnark/std/algebra/emulated/sw_emulated"
> 	"github.com/consensys/gnark/std/math/emulated"
> 	"github.com/consensys/gnark/std/math/emulated/emparams"
> 	"github.com/consensys/gnark/test"
> )
> 
> type TestBN254ScalarMul struct {
> 	Point  sw_bn254.G1Affine
> 	Scalar sw_bn254.Scalar
> 	Res    sw_bn254.G1Affine `gnark:",public"`
> }
> 
> func (c *TestBN254ScalarMul) Define(api frontend.API) error {
> 	cr, err := sw_emulated.New[emparams.BN254Fp, emparams.BN254Fr](api, sw_emulated.GetBN254Params())
> 	if err != nil {
> 		return err
> 	}
> 	res := cr.ScalarMul(&c.Point, &c.Scalar)
> 	cr.AssertIsEqual(res, &c.Res)
> 	return nil
> }
> 
> func TestMsmSolve(t *testing.T) {
> 	assert := test.NewAssert(t)
> 	x, _ := new(big.Int).SetString("1", 10)
> 	y, _ := new(big.Int).SetString("21888242871839275222246405745257275088696311157297823662689037894645226208581", 10)
> 	scalar, _ := new(big.Int).SetString("21147276235438245106538451154094232271190030085887596632745409482267565260819", 10)
> 
> 	var point = bn254.G1Affine{
> 		X: *new(fp.Element).SetBigInt(x),
> 		Y: *new(fp.Element).SetBigInt(y),
> 	}
> 	assert.True(point.IsOnCurve())
> 	var res1 bn254.G1Affine
> 	res1.ScalarMultiplication(&point, scalar)
> 
> 	witnessCircuit := TestBN254ScalarMul{
> 		Point: sw_bn254.G1Affine{
> 			X: emulated.ValueOf[emulated.BN254Fp](x),
> 			Y: emulated.ValueOf[emulated.BN254Fp](y),
> 		},
> 		Scalar: emulated.ValueOf[emulated.BN254Fr](scalar),
> 		Res: sw_bn254.G1Affine{
> 			X: emulated.ValueOf[emulated.BN254Fp](res1.X),
> 			Y: emulated.ValueOf[emulated.BN254Fp](res1.Y),
> 		},
> 	}
> 	circuit := TestBN254ScalarMul{}
> 
> 	assert.CheckCircuit(&circuit, test.WithValidAssignment(&witnessCircuit), test.WithBackends(backend.GROTH16), test.WithCurves(ecc.BN254))
> }
> ```
> 
> We don't run solidity checker by default, to ensure it is run, use the `solccheck` build tag:
> 
> ```shell
> go test -timeout 10m -tags solccheck,prover_checks -run ^TestMsmSolve$ github.com/consensys/gnark/examples/issue1186 -v -count=1
> ```
> 
> For this to work, you need to have an additional utility `gnark-solidity-checker`. We have it in a separate repository to avoid huge dependency list introduced by importing `go-ethereum`. See https://github.com/consensys/gnark-solidity-checker for how to install its dependencies and then run
> 
> ```shell
> go install github.com/consensys/gnark-solidity-checker@master
> ```
> 
> In that repository you can also see how to construct the inputs for the Groth16 verifier i.e. https://github.com/Consensys/gnark-solidity-checker/blob/main/cmd/templates.go#L72-L124. We don't have a good way to export the full proof as a byte slice yet as in PLONK solidity verifier, so you need to do it a bit manually, but hopefully the example is clear. We have an open issue for that #956.
> 
> Sorry for the difficult setup, we are now mostly focused on supporting PLONK Solidity verifier, but will try to improve Groth16 verifier also as it is still quite much used by the community.

Is there any true full example to extract the proof for plonk or groth16 and could be veified successfully? I haven't execute a successful verify on sepolia. I have to say that gnark have more effective proof generation but bad development experience. It's time to release well docs to help developers, please @ivokub 

gnark-solidity-checker/blob/main/cmd/templates.go#L72-L124 is outdated. I read the proof which is generated by calling  proof.WriteRawTo(fProof), it will report error "wrong proof size" just use the proof bytes by reading from the fProof.

Plonk : 
```
  function Verify(bytes calldata proof, uint256[] calldata public_inputs) 
  public view returns(bool success)
```

Groth16 : 
```
function verifyProof(
        uint256[8] calldata proof,
        uint256[2] calldata commitments,
        uint256[2] calldata commitmentPok,
        uint256[1] calldata input
    ) public view
```

**Reply by xiangjianmeng** [reply]
For plonk, what is the composition of proof according to the proof.json like below?

```
{
    "LRO": [
        {
            "X": "5068754427898724426029771455046457188163043055121832772770221235082819984145",
            "Y": "18851729387189152644467529618601227335279796824742946508690050081054226102983"
        },
        {
            "X": "14830782145735965492483076901946445558536609459415927456399517620548508053414",
            "Y": "15580827685522299527443199256811777558228241510091352199928951948219311809230"
        },
        {
            "X": "20594494328227607595227277976864288620450537011195542954724506488213030380987",
            "Y": "21480251863156438100534567622454859776300104255332187019007919748487243533849"
        }
    ],
    "Z": {
        "X": "15321297457799948729154487806218913299336728208571125433374617548944235581930",
        "Y": "13809380603813268219123049102850974226342616792071354973626608758561321419627"
    },
    "H": [
        {
            "X": "16176276968629495625692007549001937933181347638790225846976138172406778222543",
            "Y": "8861039016742063506250018132494519487179102555670085780539506821775718383008"
        },
        {
            "X": "12282798022621613299284252922570970414145292215383669402646936668971620962436",
            "Y": "9496235448710848146653846875929128122882849456068539133360230800683871502813"
        },
        {
            "X": "7221041329061370290158577346816658976376802272060036982397060850336836966856",
            "Y": "21145378127967707665842987755380450477663281999884742601933828927879224519235"
        }
    ],
    "Bsb22Commitments": [
        {
            "X": "21354011120776161635156329271440352425230842017783895232026276815627968366156",
            "Y": "2503995220519878249754218127162398876399673389617923656916128372470934830033"
        }
    ],
    "BatchedProof": {
        "H": {
            "X": "8154560379277721005888093297933697296100483428291290067382187887497900949045",
            "Y": "1225822929342849001212552850350154701241060330608523210882176298562927594823"
        },
        "ClaimedValues": [
            "7905947230610726753372030312970737991065038453366782965017362499127553795215",
            "3341094796214245229162044328578487915534594260089342557017923592429011680058",
            "9834495509219617880931409384233532181961783535047831432634981732996597028932",
            "17728278742384692498835662028399853678109881457279100381753119889118862150427",
            "19745499365358749352055224406866564404150967534154738557460409242825271859245",
            "374427416666498463335942982632679584718156791563566894712299048820142069677",
            "4768778765442065699146342285648850757587858299058723688296394791896390992872"
        ]
    },
    "ZShiftedOpening": {
        "H": {
            "X": "10209726027734299341457378937924842479106723123692305897524484255703264356979",
            "Y": "19861356696008529767058242788023540878864246417820538079012752037646254688413"
        },
        "ClaimedValue": "13020897359934297096469603196587540228161971757640581087425581381996022245852"
    }
}
```

**Reply by ivokub** [reply]
Indeed, our docs are not up-to-date and do not cover all the steps, particularly for the Solidity verifier. See issues:
* https://github.com/Consensys/gnark/issues/1183
* https://github.com/Consensys/gnark/issues/1094
* https://github.com/Consensys/gnark/issues/956

We are currently focusing on implementing full coverage for Linea L2 and have postponed working on non-related issues, but will get back soon.


For checking Groth16 proof in Solidity, can you also try when providing the `Prove()` function with `backend.WithProverHashToFieldFunction(sha256.New())` option? The commitment computation is hard-coded with sha256 currently.

For checking PLONK proof in Solidity, the byte array can be obtained from the `Proof` object by calling `MarshalSolidity()` method. NB! to be able to call it you have to type assert that the proof object has the method:
```go
proof, err := plonk.Prove(...)
solidityproofer, ok := proof.(interface{MarshalSolidity() []byte})
...
solidityproof := solidityproofer.MarshalSolidity()
```

**Reply by xiangjianmeng** [reply]
@ivokub Thanks a lot. Have the solidity verify contract been audited by any other auditing firm?

**Reply by ivokub** [reply]
PLONK contract has been audited. We'll publish it soon. Groth16 contract has been looked at, but imo we don't have an audit report for it.
